core/admin/mailu/utils.py

```diff
@@ -117,6 +127,127 @@ def init_app(self, app):
 proxy = PrefixMiddleware()
 
 
+class OicClient:
+    "Redirects users to OpenID Provider if configured"
+
+    def __init__(self):
+        self.app = None
+        self.client = None
+        self.extension_client = None
+        self.registration_response = None
+
+    def init_app(self, app):
+        self.app = app
+        self.client = Client(client_authn_method=CLIENT_AUTHN_METHOD)
+        self.client.provider_config(app.config['OIDC_PROVIDER_INFO_URL'])
+        self.extension_client = ExtensionClient(client_authn_method=CLIENT_AUTHN_METHOD)
+        self.extension_client.provider_config(app.config['OIDC_PROVIDER_INFO_URL'])
+        info = {"client_id": app.config['OIDC_CLIENT_ID'], "client_secret": app.config['OIDC_CLIENT_SECRET'], "redirect_uris": [ "https://" + self.app.config['HOSTNAME'] + "/sso/login" ]}
+        client_reg = RegistrationResponse(**info)
+        self.client.store_registration_info(client_reg)
+        self.extension_client.store_registration_info(client_reg)
+
+    def get_redirect_url(self):
+        if not self.is_enabled():
+            return None
+        f_session["state"] = rndstr()
+        f_session["nonce"] = rndstr()
+        args = {
+            "client_id": self.client.client_id,
+            "response_type": ["code"],
+            "scope": ["openid"],
+            "nonce": f_session["nonce"],
+            "redirect_uri": "https://" + self.app.config['HOSTNAME'] + "/sso/login",
+            "state": f_session["state"]
+        }
+
+        auth_req = self.client.construct_AuthorizationRequest(request_args=args)
+        login_url = auth_req.request(self.client.authorization_endpoint)
+        return login_url
+
+    def exchange_code(self, query):
+        aresp = self.client.parse_response(AuthorizationResponse, info=query, sformat="urlencoded")
+        if not ("state" in f_session and aresp["state"] == f_session["state"]):
```

Comment:
This will always be a CSRF attack, no? If so please log something. Do we check the referer anywhere? If not, should we?

Reply:
Added warning log entry